core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitManager.java

```diff
@@ -0,0 +1,214 @@
+  }
+
+  public class CommitService implements Closeable {
+    private final ExecutorService committerService;
+    private final ConcurrentLinkedQueue<RewriteFileGroup> completedRewrites;
+    private final List<RewriteFileGroup> committedRewrites;
+    private final int rewritesPerCommit;
+    private final AtomicBoolean running = new AtomicBoolean(false);
+
+    CommitService(int rewritesPerCommit) {
+      LOG.info("Creating commit service for table {} with {} groups per commit", table, rewritesPerCommit);
+      this.rewritesPerCommit = rewritesPerCommit;
+
+      committerService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+          .setNameFormat("Committer-Service")
+          .build());
+
+      completedRewrites = Queues.newConcurrentLinkedQueue();
+      committedRewrites = Lists.newArrayList();
+    }
+
+    /**
+     * Starts a single threaded executor service for handling file group commits.
+     */
+    public void start() {
+      Preconditions.checkState(running.compareAndSet(false, true), "Rewrite Commit service already started");
+      LOG.info("Starting commit service for {}", table);
+      // Partial progress commit service
+      committerService.execute(() -> {
+        while (running.get() || completedRewrites.size() > 0) {
+          try {
+            if (completedRewrites.size() == 0) {
+              // Give other threads a chance to make progress
+              Thread.sleep(100);
+            }
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Interrupted while processing commits", e);
+          }
+
+          // Either we have a full commit group, or we have completed writing and need to commit what is left over
+          if (completedRewrites.size() >= rewritesPerCommit || (!running.get() && completedRewrites.size() > 0)) {
+            Set<RewriteFileGroup> batch = Sets.newHashSetWithExpectedSize(rewritesPerCommit);
+            for (int i = 0; i < rewritesPerCommit && !completedRewrites.isEmpty(); i++) {
+              batch.add(completedRewrites.poll());
+            }
+
+            try {
+              commitOrClean(batch);
+              committedRewrites.addAll(batch);
+            } catch (Exception e) {
+              LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+            }
+          }
+        }
+      });
+    }
+
+    /**
+     * Places a file group in the queue to be asynchronously committed either when the queue has enough elements
+     * to do a batch of size {@link #rewritesPerCommit} or the service has been closed.
+     * @param group file group to eventually be committed
+     */
+    public void offer(RewriteFileGroup group) {
+      LOG.debug("Offered to commit service: {}", group);
+      Preconditions.checkState(running.get(), "Cannot add rewrites to a service which has already been closed");
+      completedRewrites.add(group);
```

Comment:
I think there's a race condition that could cause a group to get ignored:
* One thread enters this method and `running.get()` is true, but is then paused before adding the group.
* Another thread calls close, setting `running` to false.
* The service thread consumes the rest of `completedRewrites` and commits

I don't think this is likely to happen given the current use of this class. Plus, the commit in the service thread would take forever and the original thread calling `offer` would almost certainly run. So maybe it isn't a real problem, but it seems like it would be a simple fix to use `synchronized` block to guard `completedRewrites`.

Reply:
Yeah there is a race here which is why I put the precondition check in the bottom of the commit close function. Since this is internal (ish) I didn't want to add in a synchronize block when because any calls of "offer" after "close" are an error in our implementation. The precondition is more of a warning for us writing the code that the implementation of RewriteDatafilesAction is wrong. If we change the packaging of this we could make all of these package private :(